pgwire: cancel ongoing txns after drainnMaxWait when draining

asubiotto opened the pull request:
Instead of simply ignoring ongoing sessions that have not completed by
drainMaxWait, this PR cancels them and waits up to a timeout for this
cancellation to take effect.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/13274)
<!-- Reviewable:end -->

Commit:
pgwire: cancel ongoing txns after drainnMaxWait when draining

Instead of simply ignoring ongoing sessions that have not completed by
drainMaxWait, this PR cancels them and waits up to a timeout for this
cancellation to take effect.

asubiotto commented:
cc @andreimatei

tamird commented:
Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/server/server.go, line 528 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-Kbk3iMAYCYCgI3pDzdz:-Kbk3iMAYCYCgI3pDze-:b5ehc90) ([raw file](https://github.com/cockroachdb/cockroach/blob/3931057fa2db1a3d738deae7272913c52bd2ec54/pkg/server/server.go#L528)):*
> ```Go
> 		}
> 		netutil.FatalIfUnexpected(httpServer.ServeWith(s.stopper, pgL, func(conn net.Conn) {
> 			connCtx := log.WithLogTagStr(s.pgServer.Ctx(), "client", conn.RemoteAddr().String())
> ```

here and below are the only places that `.Ctx()` is called; instead of storing the context on the pgserver, could you create the context and the cancel function in the body of this function, and then pass the cancel function down to the pgserver constructor?

See https://github.com/cockroachdb/cockroach/issues/9807 for background on why storing contexts on structs is a bad idea.

---

*[pkg/sql/pgwire_test.go, line 254 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-Kbk3KL307SOMNpxWXIT:-Kbk3KL307SOMNpxWXIU:b-s4cky8) ([raw file](https://github.com/cockroachdb/cockroach/blob/3931057fa2db1a3d738deae7272913c52bd2ec54/pkg/sql/pgwire_test.go#L254)):*
> ```Go
> 		pgServer := s.(*server.TestServer).PGServer()
> 		{
> 			_, err := db.Begin()
> ```

this is no bueno; in Go 1.8 transactions have background goroutines, so they must _always_ be committed or aborted. ditto below.

---

*[pkg/sql/pgwire_test.go, line 259 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-Kbk6JtAy1bj-cM8BfUe:-Kbk6JtAy1bj-cM8BfUf:b-kqjah) ([raw file](https://github.com/cockroachdb/cockroach/blob/3931057fa2db1a3d738deae7272913c52bd2ec54/pkg/sql/pgwire_test.go#L259)):*
> ```Go
> 			}
> 
> 			// Set draining with no drainWait or cancelWait timeout. The
> ```

i'm not sure i "buy" this test; if the implementation of `SetDrainingImpl` is incorrect and doesn't correctly terminate client sessions, this test might continue to pass. We need to observe from the client side that the connection was or was not disrupted.

---

*[pkg/sql/pgwire_test.go, line 266 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-Kbk66AUHVrqEOEnXxWW:-Kbk66AUHVrqEOEnXxWX:b-qnj6dj) ([raw file](https://github.com/cockroachdb/cockroach/blob/3931057fa2db1a3d738deae7272913c52bd2ec54/pkg/sql/pgwire_test.go#L266)):*
> ```Go
> 			if err := pgServer.SetDrainingImpl(
> 				true,
> 				0*time.Second, /* drainWait */
> ```

you can just use `0`

---

*[pkg/sql/pgwire_test.go, line 269 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-Kbk6AbJ2h339hKVuODA:-Kbk6AbJ2h339hKVuODB:b-ilasjg) ([raw file](https://github.com/cockroachdb/cockroach/blob/3931057fa2db1a3d738deae7272913c52bd2ec54/pkg/sql/pgwire_test.go#L269)):*
> ```Go
> 				0*time.Second, /* cancelWait */
> 			); !testutils.IsError(err, "some sessions did not respond to cancellation") {
> 				t.Fatalf("unexpected error: %s", err)
> ```

%v because `err` might be nil here.

---

*[pkg/sql/pgwire/server.go, line 94 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-Kbk2owmuJCmTqVurHjq:-Kbk2owmuJCmTqVurHjr:baoj2mg) ([raw file](https://github.com/cockroachdb/cockroach/blob/3931057fa2db1a3d738deae7272913c52bd2ec54/pkg/sql/pgwire/server.go#L94)):*
> ```Go
> 		syncutil.Mutex
> 		// All session contexts are derived from this cancellable context which
> 		// in turn is derived from the server's ambient context. This context
> ```

"derived from the server's ambient context" is not a thing; ambient contexts are only used for annotation.

looks like this context is just a background context; as I mentioned in my other comment, i'd prefer if we did not store a context on this (or any) struct.

---

*[pkg/sql/pgwire/server.go, line 222 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-Kbk4HMDnGVxr9NFJcR-:-Kbk4HMDnGVxr9NFJcR0:brygna) ([raw file](https://github.com/cockroachdb/cockroach/blob/3931057fa2db1a3d738deae7272913c52bd2ec54/pkg/sql/pgwire/server.go#L222)):*
> ```Go
> }
> 
> // SetDrainingImpl is exported for testing purposes only.
> ```

create a new file, `pgwire/helpers_test.go` and put this function there; that way the function won't be available in non-test code.

---

*[pkg/sql/pgwire/server.go, line 237 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-Kbk4dkPX3sZNcIU7wVg:-Kbk4dkQS0FYhee2Sa6o:bh98ju7) ([raw file](https://github.com/cockroachdb/cockroach/blob/3931057fa2db1a3d738deae7272913c52bd2ec54/pkg/sql/pgwire/server.go#L237)):*
> ```Go
> 		s.mu.draining = drain
> 		if !drain {
> 			// If the server is returning to an operational state, the context that
> ```

this doesn't seem right. consider the scenario:
- server starts
- client connects and starts doing work in a long-running transaction
- server starts draining
- server returns to non-draining state

in this scenario the client will have its context needlessly cancelled.

I don't think we should ever be replacing the context which is handed down into client sessions; in the scenario described above, we need to abort the pending call to `cancel()` instead.

---

*[pkg/sql/pgwire/server.go, line 264 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-Kbk5kPaNDRwHV5XIb2x:-Kbk5kPaNDRwHV5XIb2y:b9qppbg) ([raw file](https://github.com/cockroachdb/cockroach/blob/3931057fa2db1a3d738deae7272913c52bd2ec54/pkg/sql/pgwire/server.go#L264)):*
> ```Go
> 	go func() {
> 		defer close(allConnsDone)
> 		for done := range connDoneChans {
> ```

iterating over `connDoneChans` needs to happen under lock; pretty sure that will deadlock given the implementation of `ServeConn`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
Review status: 0 of 5 files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/sql/pgwire/server.go, line 94 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-Kbk2owmuJCmTqVurHjq:-KbqVunCU9pUcHobYUyW:b-s2wkrm) ([raw file](https://github.com/cockroachdb/cockroach/blob/3931057fa2db1a3d738deae7272913c52bd2ec54/pkg/sql/pgwire/server.go#L94)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

"derived from the server's ambient context" is not a thing; ambient contexts are only used for annotation.

looks like this context is just a background context; as I mentioned in my other comment, i'd prefer if we did not store a context on this (or any) struct.
</blockquote></details>

Changed to not store a context.

---

*[pkg/sql/pgwire/server.go, line 222 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-Kbk4HMDnGVxr9NFJcR-:-KbuSH2H5AX3k30nUckK:bs6g1ji) ([raw file](https://github.com/cockroachdb/cockroach/blob/3931057fa2db1a3d738deae7272913c52bd2ec54/pkg/sql/pgwire/server.go#L222)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

create a new file, `pgwire/helpers_test.go` and put this function there; that way the function won't be available in non-test code.
</blockquote></details>

Done. Moved `pkg/sql/pgwire_test.go` to `pkg/sql/pgwire/pgwire_test.go` under the `pgwire_test` package as well.

---

*[pkg/sql/pgwire/server.go, line 237 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-Kbk4dkPX3sZNcIU7wVg:-KbuwX1FJArscL9g2I-N:b-et575b) ([raw file](https://github.com/cockroachdb/cockroach/blob/3931057fa2db1a3d738deae7272913c52bd2ec54/pkg/sql/pgwire/server.go#L237)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this doesn't seem right. consider the scenario:
- server starts
- client connects and starts doing work in a long-running transaction
- server starts draining
- server returns to non-draining state

in this scenario the client will have its context needlessly cancelled.

I don't think we should ever be replacing the context which is handed down into client sessions; in the scenario described above, we need to abort the pending call to `cancel()` instead.
</blockquote></details>

Added a check for draining right before cancellation.

---

*[pkg/sql/pgwire/server.go, line 264 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-Kbk5kPaNDRwHV5XIb2x:-Kbp1nH1WZJopmhhUEAN:b-rmcqfx) ([raw file](https://github.com/cockroachdb/cockroach/blob/3931057fa2db1a3d738deae7272913c52bd2ec54/pkg/sql/pgwire/server.go#L264)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

iterating over `connDoneChans` needs to happen under lock; pretty sure that will deadlock given the implementation of `ServeConn`.
</blockquote></details>

Yeah, it would deadlock. Isn't it enough to make a copy of `s.mu.connDoneChans` under lock at the point that draining is set and use that here? Nothing will be added to the original if we stay in draining mode and any connections that are added to the original will be picked up by a subsequent call to draining. Added to a comment at the top of this function to try to explain it more clearly.

---

*[pkg/server/server.go, line 528 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-Kbk3iMAYCYCgI3pDzdz:-KbuvmBuoPrajofdLY2c:biewpld) ([raw file](https://github.com/cockroachdb/cockroach/blob/3931057fa2db1a3d738deae7272913c52bd2ec54/pkg/server/server.go#L528)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

here and below are the only places that `.Ctx()` is called; instead of storing the context on the pgserver, could you create the context and the cancel function in the body of this function, and then pass the cancel function down to the pgserver constructor?

See https://github.com/cockroachdb/cockroach/issues/9807 for background on why storing contexts on structs is a bad idea.
</blockquote></details>

We had an offline discussion in which we discussed the need to keep a context around in the `pgwire.Server` to be able to use a single context for cancellation and reset it when coming out of draining mode. To avoid the need to store a context, we decided to cancel all connection contexts individually.

---

*[pkg/sql/pgwire_test.go, line 254 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-Kbk3KL307SOMNpxWXIT:-KbqW25krcMeV2l3VZu-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3931057fa2db1a3d738deae7272913c52bd2ec54/pkg/sql/pgwire_test.go#L254)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this is no bueno; in Go 1.8 transactions have background goroutines, so they must _always_ be committed or aborted. ditto below.
</blockquote></details>

Done.

---

*[pkg/sql/pgwire_test.go, line 259 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-Kbk6JtAy1bj-cM8BfUe:-KbuY5M0qx1AKF1Nx7JC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3931057fa2db1a3d738deae7272913c52bd2ec54/pkg/sql/pgwire_test.go#L259)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

i'm not sure i "buy" this test; if the implementation of `SetDrainingImpl` is incorrect and doesn't correctly terminate client sessions, this test might continue to pass. We need to observe from the client side that the connection was or was not disrupted.
</blockquote></details>

Done.

---

*[pkg/sql/pgwire_test.go, line 266 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-Kbk66AUHVrqEOEnXxWW:-KbpAz_15qIYnqvY9xv7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3931057fa2db1a3d738deae7272913c52bd2ec54/pkg/sql/pgwire_test.go#L266)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

you can just use `0`
</blockquote></details>

Done.

---

*[pkg/sql/pgwire_test.go, line 269 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-Kbk6AbJ2h339hKVuODA:-KbpAzF9AqwtxCPnS2U4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3931057fa2db1a3d738deae7272913c52bd2ec54/pkg/sql/pgwire_test.go#L269)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

%v because `err` might be nil here.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274)*
<!-- Sent from Reviewable.io -->

tamird commented:
Looking good!

---

Reviewed 7 of 7 files at r2.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[pkg/sql/pgwire/main_test.go, line 30 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-KbuxeqIEJobSzqo1owH:-KbuxeqIEJobSzqo1owI:b-mxsg1m) ([raw file](https://github.com/cockroachdb/cockroach/blob/633aa40fb880548e157092727c86a94110d89efe/pkg/sql/pgwire/main_test.go#L30)):*
> ```Go
> 	"github.com/cockroachdb/cockroach/pkg/util/randutil"
> )
> ```

add `//go:generate ../util/leaktest/add-leaktest.sh *_test.go` while you're here (checking that the path is correct). seel sql/main_test.go for reference.

---

*[pkg/sql/pgwire/pgwire_test.go, line 270 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-KbuyBVbnF9b07irg8x5:-KbuyBVbnF9b07irg8x6:b-i0f640) ([raw file](https://github.com/cockroachdb/cockroach/blob/633aa40fb880548e157092727c86a94110d89efe/pkg/sql/pgwire/pgwire_test.go#L270)):*
> ```Go
> 
> 			// Make sure that the connection was disrupted.
> 			testutils.SucceedsSoon(t, func() error {
> ```

is this retry loop needed? comment why.

---

*[pkg/sql/pgwire/pgwire_test.go, line 272 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-KbuxzsOyCsSvNOzjJcn:-KbuxzsOyCsSvNOzjJco:b-gpi6tp) ([raw file](https://github.com/cockroachdb/cockroach/blob/633aa40fb880548e157092727c86a94110d89efe/pkg/sql/pgwire/pgwire_test.go#L272)):*
> ```Go
> 			testutils.SucceedsSoon(t, func() error {
> 				if _, err := txn.Exec("SELECT 1"); err != driver.ErrBadConn {
> 					return fmt.Errorf("unexpected error: %v", err)
> ```

nit: use errors.Errorf unless the prevailing convention in this file is to use fmt

---

*[pkg/sql/pgwire/pgwire_test.go, line 295 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-KbuyWLXGeYuIBGvuC4Q:-KbuyWLXGeYuIBGvuC4R:b-vbw5jr) ([raw file](https://github.com/cockroachdb/cockroach/blob/633aa40fb880548e157092727c86a94110d89efe/pkg/sql/pgwire/pgwire_test.go#L295)):*
> ```Go
> 			// cancel any ongoing sessions and wait for 1s for the cancellation
> 			// to take effect. Note we validate that a session finishes
> 			// correctly by ensuring that SetDrainingImpl returns no error.
> ```

i think this comment is no longer accurate

---

*[pkg/sql/pgwire/pgwire_test.go, line 302 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-KbuyYBzuQWn2_U_fREM:-KbuyYBzuQWn2_U_fREN:bk608tm) ([raw file](https://github.com/cockroachdb/cockroach/blob/633aa40fb880548e157092727c86a94110d89efe/pkg/sql/pgwire/pgwire_test.go#L302)):*
> ```Go
> 			}
> 
> 			if err := pgServer.SetDraining(false); err != nil {
> ```

what's this for? how come it's before the commit, while the above case has it after the commit?

---

*[pkg/sql/pgwire/server.go, line 264 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-Kbk5kPaNDRwHV5XIb2x:-KbuzmnDFBpPjtJd7w4D:b-65lugg) ([raw file](https://github.com/cockroachdb/cockroach/blob/3931057fa2db1a3d738deae7272913c52bd2ec54/pkg/sql/pgwire/server.go#L264)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Yeah, it would deadlock. Isn't it enough to make a copy of `s.mu.connDoneChans` under lock at the point that draining is set and use that here? Nothing will be added to the original if we stay in draining mode and any connections that are added to the original will be picked up by a subsequent call to draining. Added to a comment at the top of this function to try to explain it more clearly.
</blockquote></details>

Nice. Comment looks good.

---

*[pkg/sql/pgwire/server.go, line 235 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-KbuzUw25HkGpA9N7E-y:-KbuzUw25HkGpA9N7E-z:b-7u70sj) ([raw file](https://github.com/cockroachdb/cockroach/blob/633aa40fb880548e157092727c86a94110d89efe/pkg/sql/pgwire/server.go#L235)):*
> ```Go
> 		return connCancelMap
> 	}()
> 	if connCancelMap == nil {
> ```

you should check `len == 0` here, since that's what you really care about

---

*[pkg/sql/pgwire/server.go, line 293 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-Kbv0kllQdnJ6_SeEUEx:-Kbv0klmun_CZCvaipBH:b3z8ru3) ([raw file](https://github.com/cockroachdb/cockroach/blob/633aa40fb880548e157092727c86a94110d89efe/pkg/sql/pgwire/server.go#L293)):*
> ```Go
> func (s *Server) ServeConn(ctx context.Context, conn net.Conn) error {
> 	var draining bool
> 	{
> ```

drive by, but why do we have this extra scope?

---

*[pkg/sql/pgwire/server.go, line 298 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-Kbv12-I-jmVZJLgOxC7:-Kbv12-I-jmVZJLgOxC8:b-5ijbsl) ([raw file](https://github.com/cockroachdb/cockroach/blob/633aa40fb880548e157092727c86a94110d89efe/pkg/sql/pgwire/server.go#L298)):*
> ```Go
> 		if !draining {
> 			var cancel context.CancelFunc
> 			ctx, cancel = context.WithCancel(ctx)
> ```

should also `cancel()` in the defer below.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-:-Kbv1BO3W2vPOH4Ol0bZ:br2sbr6)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/sql/pgwire/main_test.go, line 30 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-KbuxeqIEJobSzqo1owH:-KbvCkRRPM53ZTSg-wOC:blb2vgi) ([raw file](https://github.com/cockroachdb/cockroach/blob/633aa40fb880548e157092727c86a94110d89efe/pkg/sql/pgwire/main_test.go#L30)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

add `//go:generate ../util/leaktest/add-leaktest.sh *_test.go` while you're here (checking that the path is correct). seel sql/main_test.go for reference.
</blockquote></details>

This line is already at the bottom of the file.

---

*[pkg/sql/pgwire/pgwire_test.go, line 270 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-KbuyBVbnF9b07irg8x5:-KbvCruFADJKCdzGNj9j:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/633aa40fb880548e157092727c86a94110d89efe/pkg/sql/pgwire/pgwire_test.go#L270)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

is this retry loop needed? comment why.
</blockquote></details>

Done.

---

*[pkg/sql/pgwire/pgwire_test.go, line 272 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-KbuxzsOyCsSvNOzjJcn:-KbvEXUg4wy1hSnwG1cy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/633aa40fb880548e157092727c86a94110d89efe/pkg/sql/pgwire/pgwire_test.go#L272)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: use errors.Errorf unless the prevailing convention in this file is to use fmt
</blockquote></details>

Done.

---

*[pkg/sql/pgwire/pgwire_test.go, line 295 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-KbuyWLXGeYuIBGvuC4Q:-KbvEYp9_EJNT4cQgQRv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/633aa40fb880548e157092727c86a94110d89efe/pkg/sql/pgwire/pgwire_test.go#L295)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

i think this comment is no longer accurate
</blockquote></details>

Done.

---

*[pkg/sql/pgwire/pgwire_test.go, line 302 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-KbuyYBzuQWn2_U_fREM:-KbvElBJsjMLWxkhmjJH:b-5p5e3z) ([raw file](https://github.com/cockroachdb/cockroach/blob/633aa40fb880548e157092727c86a94110d89efe/pkg/sql/pgwire/pgwire_test.go#L302)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

what's this for? how come it's before the commit, while the above case has it after the commit?
</blockquote></details>

Just makes sure that there is no error if we switch back to normal operations. The ordering of this and the commit doesn't really matter and I didn't notice the two were inconsistent. I switched it to be after the commit for consistency.

---

*[pkg/sql/pgwire/server.go, line 235 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-KbuzUw25HkGpA9N7E-y:-KbvLNcp0naSqxhjv-6b:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/633aa40fb880548e157092727c86a94110d89efe/pkg/sql/pgwire/server.go#L235)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

you should check `len == 0` here, since that's what you really care about
</blockquote></details>

Done.

---

*[pkg/sql/pgwire/server.go, line 293 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-Kbv0kllQdnJ6_SeEUEx:-KbvNIQj6UVCtTpGl6_C:b4t9huf) ([raw file](https://github.com/cockroachdb/cockroach/blob/633aa40fb880548e157092727c86a94110d89efe/pkg/sql/pgwire/server.go#L293)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

drive by, but why do we have this extra scope?
</blockquote></details>

I don't know. Removed.

---

*[pkg/sql/pgwire/server.go, line 298 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-Kbv12-I-jmVZJLgOxC7:-KbvNNOwx4paytNxC4St:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/633aa40fb880548e157092727c86a94110d89efe/pkg/sql/pgwire/server.go#L298)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

should also `cancel()` in the defer below.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

nice work

---

Reviewed 1 of 7 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/sql/pgwire/pgwire_test.go, line 252 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-KbvWPs_yud30ucHGkC5:-KbvWPsaYiSop3aUpITX:br1wpr0) ([raw file](https://github.com/cockroachdb/cockroach/blob/d297ac67572b7cc59d1dbd33a2a2714287b328db/pkg/sql/pgwire/pgwire_test.go#L252)):*
> ```Go
> 	{
> 		pgServer := s.(*server.TestServer).PGServer()
> 		{
> ```

would you mind making each of the new tests a subtest? It'll give you an opportunity to give them descriptive names.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-:-KbvWZzOIh-5HweSauf3:b-iil2jj)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
Review status: 4 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/sql/pgwire/pgwire_test.go, line 252 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-KbvWPs_yud30ucHGkC5:-KbvfbSJQwGQwHRxFoDn:bt4ox1i) ([raw file](https://github.com/cockroachdb/cockroach/blob/d297ac67572b7cc59d1dbd33a2a2714287b328db/pkg/sql/pgwire/pgwire_test.go#L252)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

would you mind making each of the new tests a subtest? It'll give you an opportunity to give them descriptive names.
</blockquote></details>

Done. Thanks for the suggestion.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 7 files at r2, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> sweet

---

Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/sql/pgwire/pgwire_test.go, line 252 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-KbzmFhvGqehNRcQRBBA:-KbzmFhvGqehNRcQRBBB:byqufgb) ([raw file](https://github.com/cockroachdb/cockroach/blob/75c17d934255bb5f4615d31ec606e76501d09709/pkg/sql/pgwire/pgwire_test.go#L252)):*
> ```Go
> 	{
> 		pgServer := s.(*server.TestServer).PGServer()
> 		t.Run("TestConnCancellationResponseFailure", func(t *testing.T) {
> ```

can these be independent test functions? This test is now hard to read, with a main test and some subtests...

---

*[pkg/sql/pgwire/pgwire_test.go, line 274 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-KbzogxgJTiBQdTU-MxI:-KbzogxgJTiBQdTU-MxJ:bmmc29w) ([raw file](https://github.com/cockroachdb/cockroach/blob/75c17d934255bb5f4615d31ec606e76501d09709/pkg/sql/pgwire/pgwire_test.go#L274)):*
> ```Go
> 			// itself.
> 			testutils.SucceedsSoon(t, func() error {
> 				if _, err := txn.Exec("SELECT 1"); err != driver.ErrBadConn {
> ```

would you mind adding a top-level comment to `SetDrain` explaining where this `ErrBadConn` will come from on connections that send commands in future transactions?

---

*[pkg/sql/pgwire/pgwire_test.go, line 298 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-Kbzp5aJHKX76K1fQfcU:-Kbzp5aKS1bFgv5es7kt:bhwvk2y) ([raw file](https://github.com/cockroachdb/cockroach/blob/75c17d934255bb5f4615d31ec606e76501d09709/pkg/sql/pgwire/pgwire_test.go#L298)):*
> ```Go
> 			// cancel any ongoing sessions and wait for 1s for the cancellation
> 			// to take effect.
> 			if err := pgServer.SetDrainingImpl(
> ```

I'm confused; what's the different between this and the previous tests? I see that this test also opens a transaction. How does the cancellation take effect in this case?

---

*[pkg/sql/pgwire/server.go, line 97 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-Kbzh0itu5Eq7RxMrcBO:-Kbzh0itu5Eq7RxMrcBP:b-rqd9ed) ([raw file](https://github.com/cockroachdb/cockroach/blob/75c17d934255bb5f4615d31ec606e76501d09709/pkg/sql/pgwire/server.go#L97)):*
> ```Go
> 	mu struct {
> 		syncutil.Mutex
> 		connCancelMap cancelChanMap
> ```

can you clarify the comment on either this member of the type definition a bit? Say that it contains a the "Done" channel for each active connection and the function to cancel the session's ctx, which is supposed to eventually lead to the closing the channel.
I would have found it clearer if the map contained `SessionInfo` structs with both a channel and the CancelFunc. Also, I would replace the map with a slice, since what you want is a set of pairs, not something you can do lookups on. Right? At the very least, move away from the `Map` name.

---

*[pkg/sql/pgwire/server.go, line 200 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-Kbzl3cph-J4F3KpyGG9:-Kbzl3cph-J4F3KpyGGA:b314g4x) ([raw file](https://github.com/cockroachdb/cockroach/blob/75c17d934255bb5f4615d31ec606e76501d09709/pkg/sql/pgwire/server.go#L200)):*
> ```Go
> // served and waits a reasonable amount of time for open connections to
> // terminate. If an error is returned, the server remains in draining state,
> // though open connections may continue to exist.
> ```

I think "*though* open conns may continue to exist" is confusing, leading one to believe that it's unusual for server to be in a draining state while there are open connections.
And I think we should spell out that an error will be returned if some connections couldn't be closed.

---

*[pkg/sql/pgwire/server.go, line 217 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-KbzgCB79Pk0eHPeNl-j:-KbzgCB79Pk0eHPeNl-k:b-p19rex) ([raw file](https://github.com/cockroachdb/cockroach/blob/75c17d934255bb5f4615d31ec606e76501d09709/pkg/sql/pgwire/server.go#L217)):*
> ```Go
> 	// will be cancelled.
> 	// 2) If new connections are added to the original map, it follows that they
> 	// were added when s.mu.draining = false, thus not requiring cancellation.
> ```

what does it mean for a conn to be added when `s.mu.draining == false` ?

---

*[pkg/sql/pgwire/server.go, line 309 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-KbzkDaeKBxxOJz9zndn:-KbzkDaeKBxxOJz9zndo:b-qhw31i) ([raw file](https://github.com/cockroachdb/cockroach/blob/75c17d934255bb5f4615d31ec606e76501d09709/pkg/sql/pgwire/server.go#L309)):*
> ```Go
> 	s.mu.Unlock()
> 
> 	// If the Server is draining, we will use the connection only to send an
> ```

how about we don't add the conn to `connCancelMap` if we're draining? And then you don't have to explain in `setDraining` that you might have conns in there that don't need cancellation.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-:-KbzpM7aRHZI9QjSVUCf:b-7vyvj3)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/sql/pgwire/pgwire_test.go, line 274 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-KbzogxgJTiBQdTU-MxI:-KbzpfX0bxww61JI4DSa:bz7l0ob) ([raw file](https://github.com/cockroachdb/cockroach/blob/75c17d934255bb5f4615d31ec606e76501d09709/pkg/sql/pgwire/pgwire_test.go#L274)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

would you mind adding a top-level comment to `SetDrain` explaining where this `ErrBadConn` will come from on connections that send commands in future transactions?
</blockquote></details>

ErrBadConn is a database/sql thing; is it wise to discuss those details?

---

*[pkg/sql/pgwire/pgwire_test.go, line 298 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-Kbzp5aJHKX76K1fQfcU:-KbzpljUJlzu-0czzVjp:bmg7km4) ([raw file](https://github.com/cockroachdb/cockroach/blob/75c17d934255bb5f4615d31ec606e76501d09709/pkg/sql/pgwire/pgwire_test.go#L298)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I'm confused; what's the different between this and the previous tests? I see that this test also opens a transaction. How does the cancellation take effect in this case?
</blockquote></details>

The timeouts are different; the other test doesn't have a timeout, which is why it needs a retry loop on to observe the connection getting disrupted.

---

*[pkg/sql/pgwire/server.go, line 97 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-Kbzh0itu5Eq7RxMrcBO:-Kbzq8iVTuL6a8edbR72:b75o62u) ([raw file](https://github.com/cockroachdb/cockroach/blob/75c17d934255bb5f4615d31ec606e76501d09709/pkg/sql/pgwire/server.go#L97)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

can you clarify the comment on either this member of the type definition a bit? Say that it contains a the "Done" channel for each active connection and the function to cancel the session's ctx, which is supposed to eventually lead to the closing the channel.
I would have found it clearer if the map contained `SessionInfo` structs with both a channel and the CancelFunc. Also, I would replace the map with a slice, since what you want is a set of pairs, not something you can do lookups on. Right? At the very least, move away from the `Map` name.
</blockquote></details>

I think the point of a map is that you can avoid unbounded memory growth, which can happen with a slice.

---

*[pkg/sql/pgwire/server.go, line 309 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-KbzkDaeKBxxOJz9zndn:-KbzqLVon0gW4z5gXWhV:b-oow3gh) ([raw file](https://github.com/cockroachdb/cockroach/blob/75c17d934255bb5f4615d31ec606e76501d09709/pkg/sql/pgwire/server.go#L309)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

how about we don't add the conn to `connCancelMap` if we're draining? And then you don't have to explain in `setDraining` that you might have conns in there that don't need cancellation.
</blockquote></details>

I don't think you're understanding the comment on setdraining; that comment means that if draining is turned on and then off, any connections added after draining is turned off don't need to be closed (or even tracked for closing). And the point of that is to not hold the lock through the draining sequence.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/sql/pgwire/pgwire_test.go, line 226 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-Kbzt9LliUkpIsuwmtFS:-Kbzt9LliUkpIsuwmtFT:ble6lyu) ([raw file](https://github.com/cockroachdb/cockroach/blob/75c17d934255bb5f4615d31ec606e76501d09709/pkg/sql/pgwire/pgwire_test.go#L226)):*
> ```Go
> 		// Ensure server is in draining mode and rejects new connections.
> 		testutils.SucceedsSoon(t, func() error {
> 			if err := trivialQuery(pgBaseURL); !testutils.IsError(err, pgwire.ErrDraining) {
> ```

do we have a test for `ErrDraining` being returned on an existing connection after the transaction that was ongoing when draining started is finished?

---

*[pkg/sql/pgwire/pgwire_test.go, line 274 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-KbzogxgJTiBQdTU-MxI:-KbzvmXxYlHiwFVzkZfL:b-sz2hz9) ([raw file](https://github.com/cockroachdb/cockroach/blob/75c17d934255bb5f4615d31ec606e76501d09709/pkg/sql/pgwire/pgwire_test.go#L274)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

ErrBadConn is a database/sql thing; is it wise to discuss those details?
</blockquote></details>

well right now I'm not sure where this error is coming from. And generally the draining code has become complex (necessarily so); I think it'd be very useful to hang a big comment from `SetDraining()` explaining what's going to happen with different connections and when (and when they're going to stop rejecting new statements and why) depending on their state - if they're in a transaction versus not, if they're in the middle of a statement versus blocked on a read, if the statement responds to cancellation versus not.

---

*[pkg/sql/pgwire/pgwire_test.go, line 298 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-Kbzp5aJHKX76K1fQfcU:-KbzrzfD2q78opqq_bRw:b1uh90x) ([raw file](https://github.com/cockroachdb/cockroach/blob/75c17d934255bb5f4615d31ec606e76501d09709/pkg/sql/pgwire/pgwire_test.go#L298)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

The timeouts are different; the other test doesn't have a timeout, which is why it needs a retry loop on to observe the connection getting disrupted.
</blockquote></details>

OK, I missed the fact that these tests are not tied to the txn state at all (the 0 drainWait makes the txns irrelevant). So the difference between the tests is whether `pgServer.SetDrainingImpl()` sees the "context deadline exceeded" error returned by `v3Conn.serve()`, right?
I think some comments would be useful. And I think the transactions in these tests are misleading. Can we get rid of them and just open a connection in a pool of size one? Or... a `pgx` connection...

---

*[pkg/sql/pgwire/server.go, line 97 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-Kbzh0itu5Eq7RxMrcBO:-KbzuQ58e06VX1eN29Dc:b-62ga6r) ([raw file](https://github.com/cockroachdb/cockroach/blob/75c17d934255bb5f4615d31ec606e76501d09709/pkg/sql/pgwire/server.go#L97)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I think the point of a map is that you can avoid unbounded memory growth, which can happen with a slice.
</blockquote></details>

what do you mean? how are we limiting memory growth with this map?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/sql/pgwire/pgwire_test.go, line 298 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-Kbzp5aJHKX76K1fQfcU:-Kc-ZRsqwMQ8IqqwwyLw:b-ak27mv) ([raw file](https://github.com/cockroachdb/cockroach/blob/75c17d934255bb5f4615d31ec606e76501d09709/pkg/sql/pgwire/pgwire_test.go#L298)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

OK, I missed the fact that these tests are not tied to the txn state at all (the 0 drainWait makes the txns irrelevant). So the difference between the tests is whether `pgServer.SetDrainingImpl()` sees the "context deadline exceeded" error returned by `v3Conn.serve()`, right?
I think some comments would be useful. And I think the transactions in these tests are misleading. Can we get rid of them and just open a connection in a pool of size one? Or... a `pgx` connection...
</blockquote></details>

I think a comment would suffice; no need for the abstraction breaking power of pgx in my opinion.

---

*[pkg/sql/pgwire/server.go, line 97 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-Kbzh0itu5Eq7RxMrcBO:-Kc-ZXXM4hkjXvxaHkjG:bnudf8f) ([raw file](https://github.com/cockroachdb/cockroach/blob/75c17d934255bb5f4615d31ec606e76501d09709/pkg/sql/pgwire/server.go#L97)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

what do you mean? how are we limiting memory growth with this map?
</blockquote></details>

Removing elements from a map allows the GC to reclaim memory, while a slice will retain its lifetime maximum size.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/sql/pgwire/pgwire_test.go, line 226 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-Kbzt9LliUkpIsuwmtFS:-Kc-4456Y_XXqHh-rVw-:b-9e0892) ([raw file](https://github.com/cockroachdb/cockroach/blob/75c17d934255bb5f4615d31ec606e76501d09709/pkg/sql/pgwire/pgwire_test.go#L226)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

do we have a test for `ErrDraining` being returned on an existing connection after the transaction that was ongoing when draining started is finished?
</blockquote></details>

Not specifically, but I think checking `ErrBadConn` to ensure that a connection is disrupted in the following tests is good enough since the server doesn't shut down and it seems pretty safe to assume the disruption would come from draining the `pgServer`. The thing is that `ErrDraining` isn't returned as a response to anything (which is what is specified according to the [postgres spec](https://www.postgresql.org/docs/current/static/protocol-flow.html#AEN112869)) so we don't really have a way of observing that `ErrDraining` was returned unless we can read directly from the connection.

---

*[pkg/sql/pgwire/pgwire_test.go, line 252 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-KbzmFhvGqehNRcQRBBA:-Kc-4Gl46sQeD2oXOL4V:b-w2a900) ([raw file](https://github.com/cockroachdb/cockroach/blob/75c17d934255bb5f4615d31ec606e76501d09709/pkg/sql/pgwire/pgwire_test.go#L252)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

can these be independent test functions? This test is now hard to read, with a main test and some subtests...
</blockquote></details>

Done, it made more sense to just split the main test and these subtests rather than make the subtests their own tests because both have to do with ongoing transactions and if the first one fails, the result of the second one isn't meaningful.

---

*[pkg/sql/pgwire/pgwire_test.go, line 274 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-KbzogxgJTiBQdTU-MxI:-Kc-4OF5RUV0q96s40gC:b-dcsl91) ([raw file](https://github.com/cockroachdb/cockroach/blob/75c17d934255bb5f4615d31ec606e76501d09709/pkg/sql/pgwire/pgwire_test.go#L274)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

well right now I'm not sure where this error is coming from. And generally the draining code has become complex (necessarily so); I think it'd be very useful to hang a big comment from `SetDraining()` explaining what's going to happen with different connections and when (and when they're going to stop rejecting new statements and why) depending on their state - if they're in a transaction versus not, if they're in the middle of a statement versus blocked on a read, if the statement responds to cancellation versus not.
</blockquote></details>

I think that all that information might be a bit out of scope for the comment on `SetDraining`. I linked to the RFC.

---

*[pkg/sql/pgwire/pgwire_test.go, line 298 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-Kbzp5aJHKX76K1fQfcU:-Kc-4fy22QM2nEqiyoXn:b-uv9okc) ([raw file](https://github.com/cockroachdb/cockroach/blob/75c17d934255bb5f4615d31ec606e76501d09709/pkg/sql/pgwire/pgwire_test.go#L298)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I think a comment would suffice; no need for the abstraction breaking power of pgx in my opinion.
</blockquote></details>

The fact that we have an open transaction on a connection means that the only way for this connection to close itself is for its context to be cancelled (since it also closes itself when the server is in draining mode and there is no ongoing txn). The behavior would be the same but since the test is about cancellation, it makes sense to keep it as a txn.

`SetDrainingImpl()` itself won't observe any error returned by `v3Conn.serve()`. In the first test we make sure that we can determine when connections have not responded to cancellation in time by removing a period of time for the server to wait for this and in the second we test normal draining procedures in which we expect a connection to respond to cancellation in time.

I added some comments to the top of the subtests, let me know if those aren't enough.

---

*[pkg/sql/pgwire/server.go, line 97 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-Kbzh0itu5Eq7RxMrcBO:-Kc-6tKSkxGiZzV9KmzN:bsy1s7r) ([raw file](https://github.com/cockroachdb/cockroach/blob/75c17d934255bb5f4615d31ec606e76501d09709/pkg/sql/pgwire/server.go#L97)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

what do you mean? how are we limiting memory growth with this map?
</blockquote></details>

I had trouble with a good comment because I didn't want to make the type comment too specific. I added a comment to the member. Does that work? I'm using a map for a couple reasons:
1) At some point in the future, we're going to add cancellation of unique connections in which case it would be useful to have a map for lookup.
2) Deletion is more efficient.
The initial idea was to have the channel and the CancelFunc in a `Session` but thought it was cleaner and simpler to handle it without adding fields unused when not draining to the struct and keep draining-related stuff at the server/conn level.

---

*[pkg/sql/pgwire/server.go, line 200 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-Kbzl3cph-J4F3KpyGG9:-Kc-9BMeTh76phuBfsTJ:bfm4kef) ([raw file](https://github.com/cockroachdb/cockroach/blob/75c17d934255bb5f4615d31ec606e76501d09709/pkg/sql/pgwire/server.go#L200)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think "*though* open conns may continue to exist" is confusing, leading one to believe that it's unusual for server to be in a draining state while there are open connections.
And I think we should spell out that an error will be returned if some connections couldn't be closed.
</blockquote></details>

Updated this comment.

---

*[pkg/sql/pgwire/server.go, line 217 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-KbzgCB79Pk0eHPeNl-j:-Kc-9D-SGGp08YQoG0DC:b-x19ylp) ([raw file](https://github.com/cockroachdb/cockroach/blob/75c17d934255bb5f4615d31ec606e76501d09709/pkg/sql/pgwire/server.go#L217)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

what does it mean for a conn to be added when `s.mu.draining == false` ?
</blockquote></details>

Conns are only added to the map when `s.mu.draining == false` i.e. the server is not draining.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 7 files at r2, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[pkg/sql/pgwire/pgwire_test.go, line 252 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-KbzmFhvGqehNRcQRBBA:-Kc-gIVZQOk7OCB3c_uk:b-pp0ww3) ([raw file](https://github.com/cockroachdb/cockroach/blob/75c17d934255bb5f4615d31ec606e76501d09709/pkg/sql/pgwire/pgwire_test.go#L252)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Done, it made more sense to just split the main test and these subtests rather than make the subtests their own tests because both have to do with ongoing transactions and if the first one fails, the result of the second one isn't meaningful.
</blockquote></details>

FYI, both sub tests will run no matter what; the result of the first is not a running criterion of the second.

---

*[pkg/sql/pgwire/pgwire_test.go, line 314 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-Kc-fTfp4sD03j3NcxuV:-Kc-fTfp4sD03j3NcxuW:b-atvlbm) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c79b20af40bd1fcc2cc592687d768582ff01973/pkg/sql/pgwire/pgwire_test.go#L314)):*
> ```Go
> 	// Make sure that a connection gets cancelled and correctly responds to this
> 	// cancellation by closing itself.
> 	t.Run("TestConnCancellationResponseSuccess", func(t *testing.T) {
> ```

Btw these names (for the subtests) are super verbose. Remember that they get printed as `TestTopLevelName/SubTestName` so this will read `TestPGWireDrainCancelOngoingTxns/TestConnCancellationResponseSuccess` which is probably not what you had in mind.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/sql/pgwire/server.go, line 217 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-KbzgCB79Pk0eHPeNl-j:-Kc02TnCxkfBelODgxCb:bezsv4h) ([raw file](https://github.com/cockroachdb/cockroach/blob/75c17d934255bb5f4615d31ec606e76501d09709/pkg/sql/pgwire/server.go#L217)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Conns are only added to the map when `s.mu.draining == false` i.e. the server is not draining.
</blockquote></details>

OK, good. But then I don't understand what `2)` is telling me....
If it's what Tamir is saying in a comment below, that it's about connections coming after we stopped draining, maybe just say that conns arriving after this call are not our responsibility.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-:-Kc03N7EsNszjquR4qIT:bnfp4nl)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
Review status: 3 of 5 files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/sql/pgwire/pgwire_test.go, line 314 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-Kc-fTfp4sD03j3NcxuV:-KcIzQuZa0kvnBZ1Famc:bh4wfg8) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c79b20af40bd1fcc2cc592687d768582ff01973/pkg/sql/pgwire/pgwire_test.go#L314)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Btw these names (for the subtests) are super verbose. Remember that they get printed as `TestTopLevelName/SubTestName` so this will read `TestPGWireDrainCancelOngoingTxns/TestConnCancellationResponseSuccess` which is probably not what you had in mind.
</blockquote></details>

Simplified the names.

---

*[pkg/sql/pgwire/server.go, line 217 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274#-KbzgCB79Pk0eHPeNl-j:-KcJ4NR-C9UcSrKxcVms:b7bjgxm) ([raw file](https://github.com/cockroachdb/cockroach/blob/75c17d934255bb5f4615d31ec606e76501d09709/pkg/sql/pgwire/server.go#L217)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

OK, good. But then I don't understand what `2)` is telling me....
If it's what Tamir is saying in a comment below, that it's about connections coming after we stopped draining, maybe just say that conns arriving after this call are not our responsibility.

</blockquote></details>

Added to `2)` to clarify.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 7 files at r2, 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13274)*
<!-- Sent from Reviewable.io -->
